sys/net/credman/credman.c

```diff
@@ -90,6 +100,246 @@ int credman_add(const credman_credential_t *credential)
+        DEBUG("credman: invalid private key version\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* point to privateKeyAlgorithm */
+    node = node->next;
+    if (!node || node->type != ASN1_TYPE_SEQUENCE || !node->length) {
+        DEBUG("credman: invalid private key algorithm identifier\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* for now only ECDSA is supported by credman */
+    /* the algorithm should be Elliptic Curve Public Key (OID 1.2.840.10045.2.1) */
+    asn1_tree *algorithm = node->child;
+    if (sizeof(ecPublicKey) != algorithm->length) {
+        DEBUG("credman: wrong private key algorithm, only ecPublicKey is supported\n");
+        return CREDMAN_INVALID;
+    }
+
+    if (memcmp(ecPublicKey, algorithm->data, sizeof(ecPublicKey)) != 0) {
+        DEBUG("credman: wrong OID for algorithm. Expected 1.2.840.10045.2.1\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* point to privateKey */
+    node = node->next;
+    if (!node || node->type != ASN1_TYPE_OCTET_STRING || !node->data || !node->length) {
+        DEBUG("credman: no private key found\n");
+        return CREDMAN_INVALID;
+    }
+
+    return credman_load_private_ecc_key(node->data, node->length, cred);
+
+}
+
+int credman_load_private_ecc_key(const uint8_t *buf, size_t buf_len, credman_credential_t *cred)
+{
+    assert(buf);
+    assert(cred);
+
+    int obj_count = der_object_count((uint8_t *)buf, buf_len);
+    if (obj_count < 0) {
+        DEBUG("credman: could not calculate the number of elements within the key\n");
+        return CREDMAN_INVALID;
+    }
+
+    asn1_tree objects[obj_count];
+    asn1_tree priv_key;
+
+    if (der_decode((uint8_t *)buf, buf_len, &priv_key, objects, obj_count) < 0) {
+        DEBUG("credman: could not parse the key\n");
+        return CREDMAN_INVALID;
+    }
+
+    /*
+     * From https://tools.ietf.org/html/rfc5915#section-3
+     *
+     * ECPrivateKey ::= SEQUENCE {
+     *      version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
+     *      privateKey     OCTET STRING,
+     *      parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
+     *      publicKey  [1] BIT STRING OPTIONAL
+     * }
+     */
+
+    /* point to version, it SHALL be 1 */
+    asn1_tree *node = priv_key.child;
+    if (!node || node->type != ASN1_TYPE_INTEGER || node->data[0] != 0x01) {
+        DEBUG("credman: invalid private key version\n");
+        return CREDMAN_INVALID;
+    }
+
+    /* point to privateKey */
+    node = node->next;
+    if (!node || node->type != ASN1_TYPE_OCTET_STRING || !node->data || !node->length) {
+        DEBUG("credman: invalid private key\n");
+        return CREDMAN_INVALID;
+    }
+
+    cred->type = CREDMAN_TYPE_ECDSA;
+    cred->params.ecdsa.public_key.x = NULL;
+    cred->params.ecdsa.public_key.y = NULL;
+
+    cred->params.ecdsa.private_key = node->data;
+
+    /* try to find a publicKey by tag */
+    while (node && node->type != 0xA1) {
```

Comment:
Where does this value comes from?

Reply:
It is a context-specific tag value in DER. It is formed by `0xA0` which denotes context-specific class ORed with the tag number (the `[1]` in the `ECPrivateKey` structure. Info about this can be found in section 8.1.2.2 of [ITU-T X.690](https://www.itu.int/itu-t/recommendations/rec.aspx?rec=12483&lang=en).
I can make it a macro to document it.

Reply:
A macro to document it would definitely be better especially since this code is not touched much and it will make it easier for the next person to understand what is done.